### PR TITLE
Pull Request guidelines

### PR DESCRIPTION
**Pull request by g-gaston**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/developer/pull-requests.md
Comment (g-gaston) [inline docs/developer/pull-requests.md:41]: @danbudris @chrisdoherty4 could you guys help me with this?
You are both very good at this so I'm not sure if we can just use a concrete example from a real PR or we want to write something ad-hoc for this doc

File: docs/developer/pull-requests.md
Comment (g-gaston) [inline docs/developer/pull-requests.md:91]: @chrisdoherty4 looking at you here since you proposed this
I was thinking one clear example could be unit tests (we don't postpone unit tests unless the code is under a feature toggle)
Anything else?

File: docs/developer/pull-requests.md
Comment (g-gaston) [inline docs/developer/pull-requests.md:47]: @danbudris @chrisdoherty4 looking at you both again since you proposed this section

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [inline docs/developer/pull-requests.md:12]: I think we need to be careful here. Its worded as if test code is exempt, entirely, from the 500 LoC recommendation. I don't think that's wise. Test code is just code. It still needs reviewing like any other piece of code. Dependent on how the tests are written, i.e. if someones tried to be clever and de-duplicate a bunch of stuff, it could be low or high cognitive load.

What do we think about making it abundantly clear with the following

> The recommended maximum lines of change is 500. Split PRs as necessary and be guided by the Single Responsibility Principle. Put yourself in the reviewers shoes and consider the cognitive load on the reviewer.
>
> For example, a PR with 1000 LoC touching 20 files could be easy to review if the changes are grammar or could be difficult to review if there are functional changes in loosely related components.

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [inline docs/developer/pull-requests.md:16]:
```suggestion
Help the reviewer build context asynchronously with a well structured PR description. A good description can serve as documentation for author and reviewer when tracking down changes in the future and for outside contributors to get a feel for what a change is in regard to.
```

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [inline docs/developer/pull-requests.md:46]:
```suggestion
Reviewing your own changes increases code quality and streamlines reviews. Add comments to your own PR to notify reviewers you've identified the problem and they don't need to comment. Draw attention from reviewers on areas of code your unsure about and ask for additional input.
```

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [reply docs/developer/pull-requests.md:91]: I think unit tests is a bit of a special case. Some people do the whole "I'll write them later" but frankly I think there's rarely, if ever, a good reason to do that. 

Beyond unit tests, its really up to people to discuss it with the reviewer. Does that resonate with you?

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [reply docs/developer/pull-requests.md:41]: Will draft something shortly.

**Review by chrisdoherty4 — COMMENTED** [review]
lgtm in general

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [inline docs/developer/pull-requests.md:45]:
```suggestion
Reviewing your own changes increases code quality and streamlines reviews. Add comments to your own PR to notify reviewers you've identified the problem and they don't need to comment. Draw attention from reviewers on areas of code you're unsure about and ask for additional input.
```

My bad

File: docs/developer/pull-requests.md
Comment (g-gaston) [reply docs/developer/pull-requests.md]: Yeah, sounds good. I was only following up on this since you mentioned before we should probably add some examples about what's ok to postpone and what's not.
I added a little bit about unit tests, let me know.

Comment (chrisdoherty4) [reply docs/developer/pull-requests.md]: lgtm

File: docs/developer/pull-requests.md
Comment (danbudris) [inline docs/developer/pull-requests.md:2]:
```suggestion
This document outlines best practices to follow when creating and reviewing PRs. These are all guidelines and not hard rules, so take them as so.
```

File: docs/developer/pull-requests.md
Comment (danbudris) [inline docs/developer/pull-requests.md:8]:
```suggestion
The recommended maximum number of lines changed is 500. Split PRs as necessary and be guided by the [Single Responsibility Principle](https://blog.cleancoder.com/uncle-bob/2014/05/08/SingleReponsibilityPrinciple.html). Put yourself in the reviewers shoes and consider the cognitive load on the reviewer.
```

File: docs/developer/pull-requests.md
Comment (danbudris) [inline docs/developer/pull-requests.md:20]:
```suggestion
4. Include relevant links to the GitHub issue, design docs, documentation, etc, as needed.

```

File: docs/developer/pull-requests.md
Comment (danbudris) [reply docs/developer/pull-requests.md:37]: a link to one of these could work, I think they're pretty good examples of mine -- might be able to find a better one if we looked harder

https://github.com/aws/eks-anywhere/pull/1175
https://github.com/aws/eks-anywhere/pull/237

File: docs/developer/pull-requests.md
Comment (danbudris) [inline docs/developer/pull-requests.md:50]:
```suggestion

You can find existing EKS Anywhere design docs [here](https://github.com/aws/eks-anywhere/tree/main/designs).
```

File: docs/developer/pull-requests.md
Comment (danbudris) [inline docs/developer/pull-requests.md:53]:
```suggestion
Most of the time, there are other folks involved in the project/feature you are working on, even if it was only during the design phase. Add all of them. If no one else has been involved or your PR is just a small fix/random improvement you came up with, use the GitHub suggested reviewers. If you are in doubt, ping the team in Slack and ask for help.
```

File: docs/developer/pull-requests.md
Comment (danbudris) [inline docs/developer/pull-requests.md:67]:
```suggestion
24 hours from the time the PR is open to the time the first review is submitted. That should be more than enough to take into account timezones and PRs open at the end of the day. It gives us at least a full working day to get to it.
```

**Review by danbudris — COMMENTED** [review]
I realize one thing we don't cover is commit size and naming. Might be worth it to call out that commit messages should be useful and clear, and ideally commits should follow the SRP as much as is possible

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [reply docs/developer/pull-requests.md:8]: > lines of change

LoC

The update reads fine also.

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:84]: Instead of "ok" you should probably write "okay" or "OK" instead.

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:89]: Same. Instead of "ok" you should probably write "okay" or "OK" instead.

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:85]: Did you mean to indent this bullet under the other bullet?

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:99]:
```suggestion
* Do I think we should be solving this problem? This one can be a bit tricky. This should probably have been brought up during release, feature, or whatever planning. If it was, and the team decided to do it, move on. The answer is already yes. If it wasn’t, be careful with how you approach it but question it.
```

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:102]:
```suggestion
* Does the code do what is supposed to do (based on the proposed solution)?
```

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:106]:
```suggestion
* Is there a better solution? Was an alternative considered? If it was, do I understand why it wasn’t selected?
```
Is that what you meant?

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:109]:
```suggestion
Note: *solution* here is used quite generically and doesn't necessarily refer only to high level design. It includes code structure, code patterns, abstractions, and so on. When there is a design doc, the high-level solution should already have been agreed upon before opening the PR. but we should still ask these questions about the solution's implementation.
```

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:111]:
```suggestion
Whether or not you push back, when the answer to these questions is no, is up to you. It requires a balance between risk, helping the team move forward and the importance of the suggested changes. In general, we recommend:
```

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:113]:
```suggestion
* Don’t stop asking questions until you understand the decision being made, even if you disagree. Try to find a balance between creating too much back and forth in the PR comments and solving everything on private discussions. When a decision is made in a private conversation, summarize it in a PR comment.
```

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:120]:
```suggestion
* When necessary, encourage folks to create follow-up tickets, so feedback and improvement ideas don't get lost.
```

File: docs/developer/pull-requests.md
Comment (chrisnegus) [inline docs/developer/pull-requests.md:122]:
```suggestion
If you get here, you are done. Approve it.
```

**Review by chrisnegus — COMMENTED** [review]
Lots of good information. Most of my comments are editorial in nature.

File: docs/developer/pull-requests.md
Comment (chrisdoherty4) [reply docs/developer/pull-requests.md:12]: I've felt the pain of poorly constructed messages and over reliance on automated squash features when needing to become a git historian, but I've never found a way to get people to adopt it.

Trying to get people to adopt a good practice for an OSS project with random contributions can be even more of a challenge.

If we do want guidelines (which I think we have to because they're tricky to enforce) I suggest regurgitating this reference (my goto): https://cbea.ms/git-commit/

Comment (micahhausler) [reply docs/developer/pull-requests.md:12]: In the past I've used the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0-beta.4/) with tooling like [clog](https://github.com/clog-tool/clog-cli) or [changelog](https://github.com/skuid/changelog) for validation and changelog generation. Not required for this review, but I certainly wouldn't be opposed to adopting something like that

Comment (chrisdoherty4) [reply docs/developer/pull-requests.md:12]: If we're going to commit to auto generated change logs then conventional commits or equivalent can be super useful - release preparation can be time consuming.

Conventional commits compliment the guidelines I posted.

Comment (g-gaston) [reply docs/developer/pull-requests.md:12]: My 2c on the topic:
Standardized commit messages are very useful and I would be in favor of adopting a standard. Conventional commits works for me but I don't have a strong preference for any standard in particular. I do think that if we do this, it should go hand by hand with some kind of automation, at least for validation.

However, in my opinion, the content of the commit messages matters more than the format itself. And I believe that's where we can provide more value: on guidelines for folks on how to write good commit messages. Maybe this is where the article Chris linked can help. By no means I think this is incompatible with a standardized format, I just think we wouldn't do a lot by only agreeing on a format without trying to improve the content of the messages.

In addition, this topic resurfaces the discussion about manually squashing commits before merge. In my opinion, things like `* fix typos` or `* pass linter` don't add anything useful to the final commit message. 

That said, since I think both topics require further discussion and they have some ramifications in our processes (automation, changing the merge process to manually squash commits...), I would like to start a separate conversation and allow this PR to move forward. Once we reach agreement on the commit message topic, we can just update this doc (it's supposed to be a living document).

@micahhausler Added the clarification about this applying only to github pull requests

File: docs/developer/pull-requests.md
Comment (g-gaston) [reply docs/developer/pull-requests.md:108]: Made this more clear

**Review by chrisdoherty4 — APPROVED** [review]

**Comment by chrisdoherty4** [comment]
lgtm

Ping me when you need an lgtm command.

**Pull request merged**
